src/state/slices/stakingDataSlice/selectors.ts

```diff
@@ -169,14 +221,122 @@ export const selectRewardsByAccountSpecifier = createDeepEqualOutputSelector(
+        )
+
+        return acc
+      },
+      {}
+    )
+
+    return rewards
+  }
+)
+
 export const selectRewardsAmountByDenom = createSelector(
   selectRewardsByAccountSpecifier,
+  selectValidatorAddress,
   selectDenom,
-  (rewardsByAccountSpecifier, denom): string => {
+  (rewardsByAccountSpecifier, validatorAddress, denom): string => {
     if (!rewardsByAccountSpecifier.length) return ''
 
     const rewards = rewardsByAccountSpecifier.find(rewards => rewards.denom === denom)
 
     return rewards?.amount || ''
   }
 )
+
+export const selectAllValidators = createDeepEqualOutputSelector(
+  selectStakingData,
+  stakingData => stakingData.byvalidator
+)
+
+export const selectSingleValidator = createSelector(
+  selectStakingData,
+  selectValidatorAddress,
+  (stakingData, validatorAddress) => {
+    return stakingData.byvalidator[validatorAddress] || null
+  }
+)
+
+export const selectNonloadedValidators = createSelector(
+  selectStakingDataByAccountSpecifier,
+  selectAllValidators,
+  (stakingData, allValidators): string[] => {
+    if (!stakingData) return []
+    const initialValidatorsAddresses = stakingData.delegations
+      ? stakingData.delegations.map(x => x.validator.address)
+      : []
+    if (stakingData.undelegations) {
+      initialValidatorsAddresses.push(...stakingData.undelegations.map(x => x.validator.address))
+    }
+
+    if (stakingData.rewards) {
+      initialValidatorsAddresses.push(...stakingData.rewards.map(x => x.validator.address))
+    }
+
+    const uniqueValidatorAddresses = [...new Set(initialValidatorsAddresses)]
+    return uniqueValidatorAddresses.filter(x => allValidators[x] === undefined)
+  }
+)
+
+export const selectStakingOpportunityDataByDenom = createDeepEqualOutputSelector(
+  selectAllDelegationsCryptoAmountByDenom,
+  selectAllUnbondingsEntriesByDenom,
+  selectAllRewardsByDenom,
+  selectAllValidators,
+  (
+    allDelegationsAmount,
+    allUndelegationsEntries,
+    allRewards,
+    allValidators
+  ): StakingOpportunity[] => {
+    const result = Object.entries(allValidators).map(([validatorAddress, { apr, moniker }]) => {
+      const delegationsAmount = allDelegationsAmount[validatorAddress] ?? '0'
+      const undelegationsAmount = get(
+        allUndelegationsEntries,
+        validatorAddress,
+        [] as chainAdapters.cosmos.UndelegationEntry[]
+      )
+        .reduce((acc: BigNumber, undelegationEntry: chainAdapters.cosmos.UndelegationEntry) => {
+          acc = acc.plus(bnOrZero(undelegationEntry.amount))
+          return acc
+        }, bnOrZero(0))
+        .toString()
+
+      const rewards = get(allRewards, validatorAddress, [] as chainAdapters.cosmos.Reward[])
+        .reduce((acc: BigNumber, rewardEntry: chainAdapters.cosmos.Reward) => {
+          acc = acc.plus(bnOrZero(rewardEntry.amount))
+          return acc
+        }, bnOrZero(0))
+        .toString()
+
+      const cryptoAmount = bnOrZero(delegationsAmount)
```

Comment:
This seems like a weird computed value, where is it used?

Reply:
The reason @chxash went with this construct is because react-table expects a collection and then uses this object to render the right `<StakingOpportunities />` data for one row - it refers to the `amount` column

Reply:
Ahh ya reading this again it makes sense. For some reason I thought it was adding delegations to undelegations. all good